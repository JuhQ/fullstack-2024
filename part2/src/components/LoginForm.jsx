import { useState } from "react"
import login from "../services/login"
import noteService from '../services/notes'

const LoginForm = ({ onSuccess, onError }) => {
    const [username, setUsername] = useState("")
    const [password, setPassword] = useState("")

    const handleLogin = async (event) => {
        event.preventDefault()

        try {
            const loginResult = await login({ username, password })
            const { token } = loginResult

            onSuccess(loginResult)
            noteService.setToken(token)

            window.localStorage.setItem("loggedNoteappUser", JSON.stringify(loginResult))
            setUsername("")
            setPassword("")
        } catch (error) {
            onError(error.message)

            setTimeout(() => {
                onError("")
            }, 5000)
        }
    }

    return (<>
        <h2>Login</h2>
        <form onSubmit={handleLogin}>
            <div>
                <input
                    type="text"
                    placeholder="Username"
                    value={username}
                    onChange={({ target }) => setUsername(target.value)}
                />
            </div>
            <div>
                <input
                    type="password"
                    placeholder="Password"
                    value={password}
                    onChange={({ target }) => setPassword(target.value)}
                />
            </div>

            <button type="submit">Log in</button>
        </form>
    </>)
}

export default LoginForm