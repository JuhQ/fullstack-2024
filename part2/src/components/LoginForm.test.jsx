import { render, screen } from '@testing-library/react'
import LoginForm from './LoginForm'
import userEvent from '@testing-library/user-event'

test('<LoginForm /> can log in', async () => {
  const user = userEvent.setup()
  const successMockHandler = vi.fn()
  const errorMockHandler = vi.fn()

  render(<LoginForm onSuccess={successMockHandler} onError={errorMockHandler} />)

  const usernameInput = screen.getByPlaceholderText('Username')
  const passwordInput = screen.getByPlaceholderText('Password')
  const button = screen.getByText('Log in')

  await user.type(usernameInput, 'tunnus')
  await user.type(passwordInput, 'salasana')
  await user.click(button)

  expect(successMockHandler.mock.calls).toHaveLength(0)
})


test('snapshot', () => {
  const result = render(<LoginForm onSuccess={() => null} onError={() => null} />)

  expect(result.container).toMatchSnapshot()
})
