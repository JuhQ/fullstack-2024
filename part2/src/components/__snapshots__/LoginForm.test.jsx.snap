// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`snapshot 1`] = `
<div>
  <h2>
    Login
  </h2>
  <form>
    <div>
      <input
        placeholder="Username"
        type="text"
        value=""
      />
    </div>
    <div>
      <input
        placeholder="Password"
        type="password"
        value=""
      />
    </div>
    <button
      type="submit"
    >
      Log in
    </button>
  </form>
</div>
`;
